volatility/framework/plugins/linux/check_creds.py

```diff
@@ -0,0 +1,72 @@
+# Volatility
+
+# This file is Copyright 2019 Volatility Foundation and licensed under the Volatility Software License 1.0
+# which is available at https://www.volatilityfoundation.org/license/vsl-v1.0
+#
+
+"""
+@author:       Matt Tressler
+@contact:      matthewtressler10@gmail.com
+"""
+
+import logging
+from typing import List
+
+from volatility.framework import interfaces, renderers, exceptions, constants, contexts, objects
+from volatility.framework.automagic import linux
+from volatility.framework.configuration import requirements
+from volatility.framework.interfaces import plugins
+from volatility.framework.layers import intel
+from volatility.plugins.linux import pslist
+
+vollog = logging.getLogger(__name__)
+
+class check_creds(interfaces.plugins.PluginInterface):
+    """Checks if any processes are sharing credential structures"""
+
+    @classmethod
+    def get_requirements(cls):
+        return [
+            requirements.TranslationLayerRequirement(name='primary',
+                                                     description='Memory layer for the kernel',
+                                                     architectures=["Intel32", "Intel64"]),
+            requirements.SymbolTableRequirement(
+                name="vmlinux", description="Linux kernel symbols"),
+
+            requirements.PluginRequirement(
+                name='pslist', plugin=pslist.PsList, version=(1, 0, 0))
+        ]
+
+    def _generator(self):
+        vmlinux = contexts.Module(
+            self.context, self.config['vmlinux'], self.config['primary'], 0)
+
+        type_task = self.context.symbol_space.get_type(self.config['vmlinux'] + constants.BANG +"task_struct")
+
+        if not type_task.has_member("cred"):
+            vollog.error("this command is not supported by this profile")
+
+        creds = {}
+
+        tasks = pslist.PsList.list_tasks(self.context,
+                                         self.config['primary'], self.config['vmlinux'])
+
+        for task in tasks:
+
+            cred_addr = task.cred.dereference().vol.offset
+
+            if not cred_addr in creds:
+                creds[cred_addr] = []
+
+            creds[cred_addr].append(task.pid)
+
+        for (_, pids) in creds.items():
+            if len(pids) > 1:
+                pid_str = ""
+                for pid in pids:
+                    pid_str = pid_str + "{0:d}, ".format(pid)
+                pid_str = pid_str[:-2]
+                yield(0, [str(pid_str)])
+
+    def run(self):
+        return renderers.TreeGrid([("PIDs", str)], self._generator())
```

Comment:
Feels like we could put more useful output here?  Just a list of PIDs doesn't relate much to anything.  Is there any information from the creds structure we could output along side these (even if just an offset?)  Not sure if that's useful if this is just to list pids that share creds, but possibly something that could help people figure out which pids share which creds?

Reply:
We could list the creds address if that makes it more consistent, but it is not really something I would expect people to pivot into volsell for. The idea of listing the pids is so people can then use pslist and other process inspection plugins to dig into the suspicious/malicious one. The legitimate kernel will never share cred structures between processes, but rootkits often will get a reference to the cred structure for pid 1 / init, and then set the cred pointer of the process they want to elevate privileges for to the cred struct of init.

If you search linux_check_creds on this blogpost then you can see how it fits into investigations:

https://volatility-labs.blogspot.com/2012/09/movp-14-average-coder-rootkit-bash.html

Reply:
Ok, so essentially the only things listed in here are malicious?  Just seems like we probably know something more useful than just the pid at this point, but if not then so be it.  Happy for you to mark this resolved if you're happy that it is...  5:)

Reply:
yea, really the only point of this plugin is to say 'these pids are sharing a cred structure'... which is a system state that the legitimate kernel would never trigger, and is a favorite trick of kernel rootkits to elevate the privileges of an attacker's shell process to root.

Reply:
So the bit that concerns me with this is that you say `sharing *a* cred structure` rather than `sharing *the* cred structure`.  If there's more than one, wouldn't it be useful to know which pids share which cred structure?  If there's only one cred structure to be shared, this makes perfect sense to me, otherwise I'd think it would make sense to differentiate them?  Happy if not, but wanted to make sure it was said...  5:)